composer/algorithms/alibi/attention_surgery_functions/utils.py

```diff
@@ -0,0 +1,179 @@
+            # Builds a function (`convert_attention`) that does model surgery any GPT2Attention modules in the model.
+
+            def convert_attention(module, module_index):
+                # Do surgery (change `module` or generate a new `module` instance to return)
+                # Note that this function can (and often should for ALiBi) depend on `max_sequence_length`
+
+                # YOUR CODE HERE
+
+                return module
+            return convert_attention
+
+    In the above example, by decorating `build_gpt2_attention_converter` (which is an instance of a `SurgeryFunctionBuilder`
+    function) with `@register_surgery_function_builder(GPT2Attention)`, the ALiBi algorithm will now apply model surgery to any
+    instances of `GPT2Attention` within the model, and will apply surgery on those instances using the `convert_attention` function
+    returned by `build_gpt2_attention_converter`.
+
+    Note that `convert_attention` follows the specific signature of a `ReplacementFunction`, which has specific arguments (for additional
+    details, see `composer/utils/module_surgery.py`). However, often the model surgery functions used by ALiBi should depend on
+    a `max_sequence_length` argument. Since this argument is not used by a `ReplacementFunction`, we make use of these
+    `SurgeryFunctionBuilder` functions so that the exact functions used in model surgery can properly refer to the `max_sequence_length`
+    argument, which is not provided until runtime (see `../alibi.py`).
+    """
+
+    def _register_module(module: Type[torch.nn.Module], func: Callable) -> None:
+        assert issubclass(module, torch.nn.Module)
+        if module in replacement_policy_mapping_builder:
+            raise ValueError(f'Module {module.__name__} already has a registered SurgeryFunctionBuilder.')
+        replacement_policy_mapping_builder[module] = func
+        return
+
+    def wrapper(func: SurgeryFunctionBuilder) -> SurgeryFunctionBuilder:
+        for module in modules:
+            _register_module(module, func)
+        return func
+
+    return wrapper
+
+
+def zero_and_freeze_expand_position_embeddings(
+    module: torch.nn.Module,
+    max_sequence_length: int,
+    position_embedding_attribute: str,
+) -> None:
+    """Replaces weights with zero tensor and prevents them from being learned further.
+
+    This is intended to be used specifically for "removing" positional embeddings.
+    """
+    try:
+        pos_embedding_module = attrgetter(position_embedding_attribute)(module)
+        old_weight = getattr(pos_embedding_module, 'weight')
+        if not isinstance(old_weight, torch.nn.Parameter):
+            raise TypeError(f'Module {module._get_name()}, position embedding {position_embedding_attribute}, '
+                            f"'weight' attribute must be of type torch.nn.Module")
+        new_weight = torch.nn.Parameter(
+            torch.zeros((max_sequence_length, old_weight.shape[1]),
```

Comment:
Unless the position embeddings are 0 (or otherwise uniform at each position) each token's embedding will still be position-dependent, which we don't want.

Zeroing and setting `required_grad = False` is the simplest way (I think) to effectively make the position embedding module a no-op.

Reply:
Makes sense, good call! This is because we don't want to override the code that sums the position embeddings, right?